File: Data Structures & Algorithms/max-water-container/submission-19.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        l, r = 0, len(heights) - 1
        maxL, maxR = heights[l], heights[r]
        res = 0

        while l < r:
            if maxL < maxR:
                total = maxL * (r - l)
                res = max(res, total)
                l += 1
                maxL = max(maxL, heights[l])
            else:
                total = maxR * (r - l)
                res = max(res, total)
                r -= 1
                maxR = max(maxR, heights[r])
        return res

File: Data Structures & Algorithms/trapping-rain-water/submission-19.cs
public class Solution
{
    public int Trap(int[] height)
    {
        if (height.Length == 0)
        {
            return 0;
        }
        var res = 0;
        var l = 0;
        var r = height.Length - 1;
        var maxL = height[l];
        var maxR = height[r];
        while (l < r)
        {
            if (maxL < maxR)
            {
                l++;
                maxL = Math.Max(maxL, height[l]);
                res += maxL - height[l];
            }
            else
            {
                r--;
                maxR = Math.Max(maxR, height[r]);
                res += maxR - height[r];
            }
        }
        return res;
    }
}
